python/python_helper_sympy.py

```diff
@@ -85,116 +122,130 @@ def __init__(self):
         self.whitelist = whitelist
 
-    def visit(self, node):
+    def visit(self, node: Any) -> None:
         if not isinstance(node, self.whitelist):
             node = get_parent_with_location(node)
             raise HasInvalidExpressionError(node.col_offset)
         return super().visit(node)
 
 class CheckFunctions(ast.NodeVisitor):
-    def __init__(self, functions):
+    def __init__(self, functions: SympyMapT) -> None:
         self.functions = functions
-    def visit_Call(self, node):
+
+    def visit_Call(self, node: Any) -> None:
         if isinstance(node.func, ast.Name):
             if node.func.id not in self.functions:
                 node = get_parent_with_location(node)
                 raise HasInvalidFunctionError(node.col_offset, node.func.id)
         self.generic_visit(node)
 
 class CheckVariables(ast.NodeVisitor):
-    def __init__(self, variables):
+    def __init__(self, variables: SympyMapT) -> None:
         self.variables = variables
-    def visit_Name(self, node):
+
+    def visit_Name(self, node: Any) -> None:
         if isinstance(node.ctx, ast.Load):
             if not is_name_of_function(node):
                 if node.id not in self.variables:
                     node = get_parent_with_location(node)
                     raise HasInvalidVariableError(node.col_offset, node.id)
         self.generic_visit(node)
 
-def is_name_of_function(node):
+def is_name_of_function(node: Any) -> bool:
     # The node is the name of a function if all of the following are true:
     # 1) it has type ast.Name
     # 2) its parent has type ast.Call
     # 3) it is not in the list of parent's args
-    return isinstance(node, ast.Name) and isinstance(node.parent, ast.Call) and (node not in node.parent.args)
+    return isinstance(node, ast.Name) and isinstance(node.parent, ast.Call) and (node not in node.parent.args) # type: ignore
 
-def get_parent_with_location(node):
+def get_parent_with_location(node: ast.AST) -> ast.AST:
     if hasattr(node, 'col_offset'):
         return node
-    else:
-        return get_parent_with_location(node.parent)
 
-def evaluate(expr, locals_for_eval={}):
+    return get_parent_with_location(node.parent) # type: ignore
+
+def evaluate(expr: str, locals_for_eval: LocalsForEval) -> sympy.Expr:
 
     # Disallow escape character
-    if '\\' in expr:
-        raise HasEscapeError(expr.find('\\'))
+    ind = expr.find('\\')
+    if ind != -1:
+        raise HasEscapeError(ind)
 
     # Disallow comment character
-    if '#' in expr:
-        raise HasCommentError(expr.find('#'))
+    ind = expr.find('#')
+    if ind != -1:
+        raise HasCommentError(ind)
 
     # Parse (convert string to AST)
     try:
         root = ast.parse(expr, mode='eval')
-    except Exception as err:
-        raise HasParseError(err.offset)
+    except SyntaxError as err:
+        offset = err.offset if err.offset is not None else -1
+        raise HasParseError(offset)
 
```

Comment:
could we miss catching certain Exceptions with this change?

Reply:
Technically yes but a generic exception isn't guaranteed to have an `offset`, so trying to access that will just throw another exception. Either way, in Python you should always avoid catching a generic exception if possible (since this will catch sys calls and stuff).